wiredtiger/wiredtiger
# WT-7832 Add an encryptor extension that uses the libsodium cryptography library.

It should really be audited by a cryptographer before being used, but
is expected to be usable with at worst minor adjustments.

It uses the aead_xchacha20poly1305_ietf construction from libsodium to
encrypt and checksum blocks.

It does not support retrieving keys from a key manager, there not
being any obvious open-source choices that I'm aware of. This means
that it can, for the time being anyway, only be configured with
secretkey= and not keyid=, which is perhaps unfortunate but better
than nothing.

Besides the encryptor itself, this changeset includes the following
related changes:
   - Add the new extension to both the cmake and autotools builds.
   - Rework the encryption page in the documentation, adding the new
     encryptor and expanding on some of the other material, and also
     add some bits/make some improvements to the WT_ENCRYPTOR docs.
   - In util_main.c, add a wt_explicit_zero function for zeroing
     memory that takes precautions against being removed by the
     compiler, and use it to clear copies of the secret key. Zero and
     free the secret key (and open config string, which contains the
     secret key when there is one) earlier.
   - In nop_encryptor.c, since this is supposed to be a template for
     application developers to fill in, add a blank customize method.
     Without a customize method you can't configure keys, so even
     though it's officially optional it seems like the example should
     have one.
   - Add support for the new extension to:
        -- test_format
        -- test_encrypt0[126].py
        -- test_import0[69].py
     (Note that test_encrypt05 doesn't exist, test_encrypt0[34] are
     for testing the config plumbing and not any particular extension,
     and test_encrypt07 needs to be able to munge the "encrypted" data
     and doesn't work with real encryption.)
   - Add new test_encrypt0[89].py that checks the error paths in the
     new extension's customize method.
   - Add an example snippet for how to configure the new extension to
     ex_all.c for use in the docs.
   - Add the encryptor directory to Doxyfile so it can be an example.
   - Add the new encryptor to the examples page in the documentation.
   - Add a bunch of spelling words.
   - Add some of the functions to the exception list in s_void (like
     other extensions).

It also includes the following change that is not related but directly
adjacent to a piece of the above:
   - In the cmake build of test_format, pass the path to the zstd
     library with -D, like the other extensions.
